### PR TITLE
Usm pool implementation

### PR DESCRIPTION
**Pull request by bmyates**

### CODE REVIEW

**Comment by bmyates** [comment]
@jandres742  This is initial usm pool implementation. It passes most CTS after applying fixes to CTS here https://github.com/oneapi-src/unified-runtime/pull/529

**Comment by jandres742** [comment]
thanks @bmyates . Seems like your patch includes all the changes. could you resubmit so the diff shows only the changes intended for this PR?

**Comment by bmyates** [comment]
It had only my changes when I opened the PR, I think url0 must have been rebased.   

I will rebase this

**Comment by bmyates** [comment]
@jandres742  ready to review

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp [hunk 1]
```diff
@@ -518,28 +528,81 @@ static ur_result_t USMAllocationMakeResident(
...
-    ur_usm_pool_handle_t Pool    ///< [in] pointer to USM memory pool
+    [[maybe_unused]] ur_context_handle_t
+        Context,              ///< [in] handle of the context object
+    ur_usm_pool_handle_t Pool ///< [in] pointer to USM memory pool
 ) {
-  std::ignore = Context;
-  std::ignore = Pool;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+
```
Comment (jandres742) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp:604]: remove the maybe_unused and add it as std::ignore for uniformity with rest of code.

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp [hunk 2]
```diff
@@ -518,28 +528,81 @@ static ur_result_t USMAllocationMakeResident(
...
+  }
+
+  HostMemPool = std::make_unique<USMAllocContext>(
+      std::unique_ptr<SystemMemory>(new USMHostMemoryAlloc(Context)),
+      this->USMAllocatorConfigs.Configs[usm_settings::MemType::Host]);
+
+  for (auto device : Context->Devices) {
+    DeviceMemPools[device] = std::make_unique<USMAllocContext>(
+        std::unique_ptr<SystemMemory>(
+            new USMDeviceMemoryAlloc(Context, device)),
```
Comment (jandres742) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp:566]: should we have try/catch for these new statements?

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp [hunk 3]
```diff
@@ -518,28 +528,81 @@ static ur_result_t USMAllocationMakeResident(
...
+        this->USMAllocatorConfigs.Configs[usm_settings::MemType::Device]);
+
+    SharedMemPools[device] = std::make_unique<USMAllocContext>(
+        std::unique_ptr<SystemMemory>(
+            new USMSharedMemoryAlloc(Context, device)),
```
Comment (jandres742) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp:571]: what happens if MaxPoolableSize doesnt fit in the device?

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp (hunk 2 above)
Comment (bmyates) [reply sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp:566]: The whole constructor is called inside a try/catch block that will catch any exceptions new throws

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp (hunk 3 above)
Comment (bmyates) [reply sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp:571]: Nothing. MaxPoolableSize is only used to determine if a requested allocation should be served from the pool or directly allocated.

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp (hunk 1 above)
Comment (bmyates) [reply sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_usm.cpp:604]: Changed, but fwiw maybe_unused is the correct way to suppress an unused variable warning in c++17.

**Comment by jandres742** [comment]
@bmyates please rebase

**Comment by bmyates** [comment]
@jandres742  rebased

**Comment by bmyates** [comment]
@jandres742  rebased again and updated to new USM pool apis

**Review by jandres742 — APPROVED** [review]

**Pull request merged**